### PR TITLE
[HttpKernel] Add a controller argument resolver for backed enums

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | N/A
| License       | MIT
| Doc PR        | Todo

Given:

```php
namespace App\Model;

enum Suit: string 
{
    case Hearts = 'H';
    case Diamonds = 'D';
    case Clubs = 'C';
    case Spades = 'S';
}
```

and the controller:

```php
class CardController
{
    #[Route('/cards/{suit}')]
    public function list(Suit $suit): Response
    {
        // [...]
    }
}
```

A request to `/cards/H` would inject the `Suit::Hearts` enum case into the controller `$suit` argument.

This core resolver aims to resolve backed enum from route path parameters, so we assume a 404 Not Found should be thrown on an invalid backing value provided.

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+                continue;
+            }
+
+            try {
+                yield $enumType::from($value);
+
+                continue;
+            } catch (\ValueError|\TypeError $error) {
+                throw new BadRequestException(sprintf('Could not resolve the "%s" controller argument: %s', $argument->getName(), $error->getMessage()), $error->getCode(), $error);
```
Comment (derrabus) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php]: I would expect a failed hydration of the enum to result in a 404 error.

Comment (ogizanagi) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php]: That's arguable. But for the main use-case targeted by this PR, i.e route parameters, it would make sense indeed.
Let's assume `\ValueError` => not found?

Since variadic would only be possible by using attributes created in the request by other means (listeners, route defaults, …), we might want to cut down the use-cases for this core resolver as well, and eventually drop out support for variadic? 
Hence any other possible issue would be a logic exception.

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            yield null;
+
+            return;
+        }
+
+        if (!\is_int($value) && !\is_string($value)) {
+            throw new \LogicException(sprintf('Could not resolve the "%s $%s" controller argument: expecting an int or string, got %s.', $argument->getType(), $argument->getName(), get_debug_type($value)));
+        }
+
+        /** @var class-string<\BackedEnum> $enumType */
```
Comment (ogizanagi) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php:58]: What should we do about this Psalm error?

```php
UndefinedDocblockClass
src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php:59:9: 
UndefinedDocblockClass: Docblock-defined class, interface or enum named BackedEnum does not exist (see https://psalm.dev/200)
```
I added an `UndefinedDocblockClass` [section in psalm.xml](https://github.com/symfony/symfony/pull/44831/commits/c5eeca7d4e17f68a926cdab9538cd3f53091c5ca) to fix it  for now (should be done in a dedicated PR targeting 4.4 if that's we want).

But would it be an issue to upgrade the Psalm workflow to use PHP 8.1?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php:58]: psalm is just a helper, it defines no goals at all. The only thing we need to do when seeing psalm errors is think about them and decide if it's applicable to the situation or not. So, what should we do? Analyze, and ignore here.

Comment (ogizanagi) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php:58]: Agreed, but at least adding adding this section to the 4.4 branch as already done for `UndefinedClass` would prevent further questions about this topic. Right? I'll open the PR.

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/BackedEnumValueResolver.php]: 👍🏻 I don't see a use-case for variadics either at the moment.

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ogizanagi.
